Do not "set" the wallpaper during startup

psychon opened the pull request:
This basically reverts 4f5e0e7: During startup, i3 would take a
screenshot and set that as the background pixmap of the root window.
This is the easy part of setting an X11 wallpaper.

The code in question was added because something either set the
background pixmap of the root window to NONE or the X11 server was
started with "-background none". I'm tempted to say "the people in
question get what they asked for", but perhaps telling them to "exec
--no-startup-id xsetroot -solid black" in their config is a better
reaction.

(I have not found any issue / bug report from "back then" which caused the code
in question to be added, so I cannot really say more.)

Fixes: https://github.com/i3/i3/issues/4371
Fixes: https://github.com/i3/i3/issues/2869
Signed-off-by: Uli Schlachter <psychon@znc.in>

Airblader commented:
It seems for years I was telling lies, saying that i3 does not manage/set the wallpaper. :-( As @stapelberg wrote in the commit message (albeit that was eight years ago *cough*)

> This commit has some chance of breakage, so I’m prepared to revert it unless we can figure out the issues and roll forward.

I'm OK with the change, but I think we'll need to document this somehow. At the very least, can you please update `RELEASE-NOTES-next` to add a note about this?

stapelberg commented:
AFAIR, this was a problem that was reported with certain login managers (lightdm perhaps? or slim?).

It might be good to check that this commit works with the most common login managers in use today.

psychon commented:
> At the very least, can you please update RELEASE-NOTES-next to add a note about this?

Sure. However, I am not quite sure about the usual file format of these files (neither of the two `RELEASE-NOTES` files has much content). What do you think about the following?

```diff
diff --git a/RELEASE-NOTES-next b/RELEASE-NOTES-next
index 928f276a..a5b4b75a 100644
--- a/RELEASE-NOTES-next
+++ b/RELEASE-NOTES-next
@@ -21,6 +21,7 @@ strongly encouraged to upgrade.
   • Add %machine placeholder (WM_CLIENT_MACHINE) to title_format
   • Allow multiple output names in 'move container|workspace to output'
   • Add 'move container|workspace to output next'
+  • During startup the wallpaper was previously set to a screenshot
 
  ┌────────────────────────────┐
  │ Bugfixes                   │
```

Alternative suggestion would be "No longer take a screenshot and set it as the wallpaper during startup", but that seems a bit long. "Stop setting the wallpaper to a screenshot during startup"?

> It might be good to check that this commit works with the most common login managers in use today.

To whoever will be testing this: Make sure your i3 config does not start anything that sets a wallpaper, since this basically defeats the purpose. As I understand it, the testing strategy would be:
- start i3
- start some window (e.g. a terminal)
- switch to another workspace

If your root window has its background pixmap set to `NONE`, the old "pixels" will stay visible: You get a terminal *without* a blinking cursor and without being able to interact with it.

(Last time I switched to another VT, Xorg no longer allowed me to switch back, so I pass. The error was something like `drmSetMaster() failed` according to `Xorg.0.log`. Sorry.)

Airblader commented:
> What do you think about the following?

The intention was never to "screenshot and copy it over"( even if that is what it did), so to avoid confusion I'd phrase it a bit more abstract, like

> Do not copy the pixmap of the root screen on startup

As for testing, we're looking at an almost fresh release cycle ahead of us, so I'd be willing to merge it without testing it extensively upfront.

psychon commented:
> Do not copy the pixmap of the root screen on startup

I went with the slightly longer "Do not copy the root window's content to its background pixmap on startup". I would like to write "to a new pixmap and use that as the root window's background pixmap", but that seems too long.

stapelberg commented:
> I went with the slightly longer "Do not copy the root window's content to its background pixmap on startup". I would like to write "to a new pixmap and use that as the root window's background pixmap", but that seems too long.

Optimizing for short lines is a secondary concern for our release notes.

Instead, I would encourage you to think from the perspective of a user. Repeatedly ask yourself “I have no idea what any of these words mean” and “but what does this mean for me?”.

In this particular case, when would people look for this commit? Likely if their background unexpectedly changes from the last version of i3 to the current version of i3. Can you update the message to briefly mention this case, so that users have something to go on when debugging?

Airblader commented:
That's why I raised the question of "where to document this". I don't expect any user to find it buried in the list of commits of the release notes, to be honest. It might rather be worth featuring this prominently in the summary of the release notes itself?

stapelberg commented:
> That's why I raised the question of "where to document this". I don't expect any user to find it buried in the list of commits of the release notes, to be honest. 

I have thus far considered the lists in the release notes to be directed at users, not merely a git commit dump. The longest bit about getting a new release out is (was?) re-writing commit messages to make sense for users.

> It might rather be worth featuring this prominently in the summary of the release notes itself?

Sounds like we need to run a few tests to gather more certainty as to whether this is a behavior-affecting change or not :)

Airblader commented:
It's not a commit dump, of course, but it is a long list of things that ultimately represent PRs/commits. It's not exactly a release blog post with shiny pictures and videos and marketing language of "what's new in i3 4.xyz". :-)

I have my doubts that most people update and carefully read the release notes (if at all), especially since we try to make compatible update paths exactly so that users don't have to worry about a changelog before updating. Of course I might be wrong.

stapelberg commented:
> I have my doubts that most people update and carefully read the release notes (if at all), especially since we try to make compatible update paths exactly so that users don't have to worry about a changelog before updating. Of course I might be wrong.

I agree with that, but at the same time, I think the release notes are the first place where people would start looking for details on what has changed if they encounter a problem after an upgrade. It’s certainly easier than to dig through actual git commits yourself :)

Airblader commented:
Right, and I'm totally on-board with phrasing it for them to find it easier. But in terms of offering a solution, we should probably feature that more permanently, or at least I think that would be quite helpful.

I've even considered whether we should add a commented-out xsetroot call in the default config that roughly says something along the lines of "if you don't use anything that sets a wallpaper, running something like this is a good idea".

psychon commented:
> I've even considered whether we should add a commented-out xsetroot call in the default config that roughly says something along the lines of "if you don't use anything that sets a wallpaper, running something like this is a good idea".

That  seems quite sensible to me. I can do that in this PR if wanted.

> Repeatedly ask yourself “I have no idea what any of these words mean” and “but what does this mean for me?”.

Okay... current version:

- Do not copy the root window's content to its background pixmap on startup

My proposal from the above:

- i3 used to take a screenshot and set it as the wallpaper when initially starting up. This was done some years ago to work around problems with some display manager, but caused problems when the screenshot was actually showing windows. Thus, this code was dropped in this release. If you do not set a wallpaper, it is recommended to add `exec --no-startup-id xsetroot -solid black` to your config.

Reason why I do not like the above: Way too verbose compared to the other lines in there.

stapelberg commented:
> * i3 used to take a screenshot and set it as the wallpaper when initially starting up. This was done some years ago to work around problems with some display manager, but caused problems when the screenshot was actually showing windows. Thus, this code was dropped in this release. If you do not set a wallpaper, it is recommended to add `exec --no-startup-id xsetroot -solid black` to your config.

This sounds good to me!

It’s the perfect length for the prose section above the lines. Depending on further testing, we should either put it into prose (if we find any breakage) or list (if we find no breakage). Either way, length is fine :)

psychon commented:
It is possible I finally understood what you meant. New version of the last commit pushed (I replaced the `RELEASE-NOTES` commit with a different one).

stapelberg commented:
The commits look good, but we still need to test the change, I think.

What does the test matrix look like (i.e. which display managers, in which configurations)? What is the test procedure?

psychon commented:
> What is the test procedure?

So far I only tested starting an `xterm` before `i3` in `Xephyr` and switching to another (empty) workspace. Before this change, I ended up with a non-working `xterm`. This is basically the test for "fixes #2869". (Oh and: I ran `i3 -c /dev/null` to make sure it does not set a wallpaper.)

With relation to display managers (=testing that this does not reintroduce https://github.com/i3/i3/commit/4f5e0e794f61cdcea2652f964ec25d0b7418f69b): If I understand the commit message correctly, this means:

- start `i3` (and make sure it does not set a wallpaper while starting)
- open some window
- switch to another (empty) workspace

If the background pixmap of the root window is set to `None`, the X11 server will not update the screen contents. Thus, you end up with a non-working "screenshot" of the window that you opened.

(Heh, before this I did not notice that both scenarios show basically the same problematic behaviour.)

> What does the test matrix look like (i.e. which display managers, in which configurations)? 

So far, I did no testing at all. (Well, only the testing I mentioned above.)

I just removed my wallpaper command from `~/.config/i3/config`, so somewhen in the next days I will be testing with lightdm.

https://github.com/i3/i3/commit/4f5e0e794f61cdcea2652f964ec25d0b7418f69b talks about GDM3. I do not have that installed and will not install it to test this, sorry.

Edit:

> I just removed my wallpaper command from ~/.config/i3/config, so somewhen in the next days I will be testing with lightdm.

I just ended up with the debian default wallpaper as my wallpaper. Nothing special happened and no breakage.

psychon commented:
Judging from https://github.com/GNOME/gdm/blob/fb319044165f2b4c6e194cf26380c29efc675f64/daemon/gdm-server.c#L221 and https://github.com/GNOME/gdm/blob/349e6af855a1c1dbee9af3e5881e00b66482d155/daemon/gdm-x-session.c#L275-L276, I would say that gdm most likely still starts the X11 server with `-background none`, so that definitely needs some testing.

Bonus points for `ps aux | grep background` and making sure the X11 server was really started with `-background none`.

Is anyone here able to easily do this test? Any volunteers?

Edit: With my current lightdm session, I get:
```
$ ps aux | grep Xorg
root         575  3.5  0.2 1003216 87672 tty7    Ssl+ 08:44   6:55 /usr/lib/xorg/Xorg :0 -seat seat0 -auth /var/run/lightdm/root/:0 -nolisten tcp vt7 -novtswitch

```

stapelberg commented:
> [4f5e0e7](https://github.com/i3/i3/commit/4f5e0e794f61cdcea2652f964ec25d0b7418f69b) talks about GDM3. I do not have that installed and will not install it to test this, sorry.

With gdm on Fedora 33, I indeed get the following Xorg command line which includes `-background none`:
```
michael         8983  20   0  0.1  0.1          \_ /usr/libexec/Xorg vt3 -displayfd 3 -auth /run/user/1000/gdm/Xauthority -nolisten tcp -background none -noreset -keeptty -novtswitch -verbose 3
```

I haven’t had time to test i3 with this pull request patched in yet.

stapelberg commented:
> If the background pixmap of the root window is set to `None`, the X11 server will not update the screen contents. Thus, you end up with a non-working "screenshot" of the window that you opened.

I just tested your pull request on Fedora 33 with gdm (default setup), and this is exactly the problem that I’m running into.

Hence, I don’t think just reverting the commit is good enough :-/.

What other options do we have?

psychon commented:
> I just tested your pull request on Fedora 33 with gdm (default setup),

Thanks for that!

> What other options do we have?

Uhm...
- Leave the status quo and accept the bugs it causes (pre-existing programs end up in the screenshot i3 creates)
- Add code to fully set the wallpaper. I do not want to go there, but did that already before for AwesomeWM, so I could do that. However, no idea where the pixels for the wallpaper should come from. Cairo can only load PNGs and that is unlikely to be enough, so would require new dependencies.
- Add code to set a black wallpaper. No idea how to tell people to enable this. Is "set wallpaper before i3 starts" a supported configuration? If not, i3 could just always set a black background.
- A variant of the above: Only set a black wallpaper (or alternatively a screenshot) if the root window does not have a `_XROOTPMAP_ID` or `ESETROOT_PMAP_ID` property. By convention, wallpaper setters are supposed to set these properties to the pixmap of the wallpaper, but since this is X11, this is only a convention. I bet `xsetroot` does not set these properties since it does not set a pixmap as background.

AFAIK there is no way to query the background of an X11 window. Meh. :-(

One could also determine the value at startup experimentally: Find an empty spot where the wallpaper would be visible. Map a window with a white background at this spot. Unmap it. Grab a "screenshot" and check if that pixel is white. (And then repeat the same with black.) Also ugly. (Edit: No window required. Just draw to the root window and use `ClearArea` to tell the X11 server to erase the drawing and re-apply the background. But this still needs an "empty space".)

By the way: `_XROOTPMAP_ID` and `ESETROOT_PMAP_ID` exists so that clients can query the wallpaper (see above: No way to query the background of a window). This is e.g. used by urxvt for fake transparency. I guess this is also currently broken with i3's screenshot since these properties are not set. Dunno if that is a bug or a feature. ;-)

Seems like the following WMs do not set a background when started (or my search for `BackPixmap` is too simplistic):
- Fluxbox: https://github.com/fluxbox/fluxbox/search?q=BackPixmap
- AwesomeWM's C code does not actually do anything unless Lua tells it to, but the default config sets a wallpaper.
- twm: https://github.com/freedesktop/twm/search?q=BackPixmap (well, okay, I am out of ideas where to search so do not take this line too seriously.)

psychon commented:
> One could also determine the value at startup experimentally: Find an empty spot where the wallpaper would be visible. Map a window with a white background at this spot. Unmap it. Grab a "screenshot" and check if that pixel is white.

I tried implementing this and tested with `Xephyr -background none`. However, it does not work. The screenshot always come back black.

However, I also cannot reproduce the original issue with `Xephyr` and this branch:

>> If the background pixmap of the root window is set to None, the X11 server will not update the screen contents. Thus, you end up with a non-working "screenshot" of the window that you opened.
>
> I just tested your pull request on Fedora 33 with gdm (default setup), and this is exactly the problem that I’m running into.

@stapelberg Can you perhaps try this program with your X11 server and without i3 / with i3 built from this PR? The desired outcome is two different values for "pixel 1" and "pixel 2", but I only ever get it to print 0.

https://gist.github.com/psychon/05d666fb123a019a636430e36c938b69

stapelberg commented:
> * Add code to set a black wallpaper. No idea how to tell people to enable this. Is "set wallpaper before i3 starts" a supported configuration? If not, i3 could just always set a black background.

Yeah, people (like me ;)) might be using `xsetroot` in their `~/.xsession` before starting i3.

I’m leaning towards auto-detecting (if possible) this issue and reacting accordingly, or, if that’s not possible, leave things at the status quo.

I ran the tests you asked on Fedora:

### Standard i3 without your pull request

When I run your `try_detect_root_background_none` from my `~/.xsession` before starting i3, I get:

```
There are 1 free/visible rectangles
Rect 0: top left 0x0 size 3840x2160
Pixel 1: 0; pixel 2: 255
```

If I run it after starting i3, I get:
```
There are 1 free/visible rectangles
Rect 0: top left 0x0 size 3840x2123
Pixel 1: 255; pixel 2: 255
```

Note that I can see a white rectangle at the top left of the screen.

### With this pull request applied

Before:
```
There are 1 free/visible rectangles
Rect 0: top left 0x0 size 3840x2160
Pixel 1: 0; pixel 2: 255
```

After:
```
There are 1 free/visible rectangles
Rect 0: top left 0x0 size 3840x2123
Pixel 1: 0; pixel 2: 255
```

White rectangle still visible.

psychon commented:
> I ran the tests you asked on Fedora:

Thanks a lot! Test result is "the test works". No idea why it does not work under `Xephyr`.

> I’m leaning towards auto-detecting (if possible)

Okay, I will update this PR accordingly, but likely not today. I will report back when I found the time.

> Note that I can see a white rectangle at the top left of the screen.

Yeah, that's basically exactly what the test does. Open a white rectangle and test if it stays white after the window was closed.

psychon committed:
Move screenshot code to libi3

This is a small refactoring. The only functional change is that I moved
the call to xcb_flush() to the end of the function so that the two
FreeXXX requests are flushed at the same time as the rest.

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon committed:
"Set" the wallpaper during startup only sometimes

Since commit 4f5e0e7, i3 would take a screenshot and set that as the
background pixmap of the root window during startup. This is the easy
part of setting a proper X11 wallpaper.

The code in question was added because something either set the
background pixmap of the root window to NONE or the X11 server was
started with "-background none". This is apparently done by default by
e.g. gdm to avoid some flickering while the X11 server starts up.

This commit makes this code conditional: Only when no wallpaper is
detected is a screenshot taken.

Since I could not find any way to query the background of a window, a
more direct approach is taken to detect this situation: First, we find
some part of the root window that is not currently covered. Then we open
a white window there, close it again and grab a screenshot. If a
wallpaper is set, the X11 server will draw this wallpaper after the
window is closed and something else will be visible in the screenshot.

However, the wallpaper could have a white pixel at the tested position.
Thus, this procedure is repeated with a black window.

Only when this procedure produces two different pixel values is a
screenshot taken and set as the wallpaper.

Fixes: https://github.com/i3/i3/issues/4371
Fixes: https://github.com/i3/i3/issues/2869
Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon commented:
I finally got around to updating this PR. It is completely redone. `git show --stat` of the result:
```
commit bf881c156f3e379e9f6f8556d9756b43d16428db
Author: Uli Schlachter <psychon@znc.in>
Date:   Tue May 18 17:19:00 2021 +0200

    "Set" the wallpaper during startup only sometimes
    
    Since commit 4f5e0e7, i3 would take a screenshot and set that as the
    background pixmap of the root window during startup. This is the easy
    part of setting a proper X11 wallpaper.
    
    The code in question was added because something either set the
    background pixmap of the root window to NONE or the X11 server was
    started with "-background none". This is apparently done by default by
    e.g. gdm to avoid some flickering while the X11 server starts up.
    
    This commit makes this code conditional: Only when no wallpaper is
    detected is a screenshot taken.
    
    Since I could not find any way to query the background of a window, a
    more direct approach is taken to detect this situation: First, we find
    some part of the root window that is not currently covered. Then we open
    a white window there, close it again and grab a screenshot. If a
    wallpaper is set, the X11 server will draw this wallpaper after the
    window is closed and something else will be visible in the screenshot.
    
    However, the wallpaper could have a white pixel at the tested position.
    Thus, this procedure is repeated with a black window.
    
    Only when this procedure produces two different pixel values is a
    screenshot taken and set as the wallpaper.
    
    Fixes: https://github.com/i3/i3/issues/4371
    Fixes: https://github.com/i3/i3/issues/2869
    Signed-off-by: Uli Schlachter <psychon@znc.in>

 RELEASE-NOTES-next        |   7 +++++++
 include/libi3.h           |   8 ++++++++
 libi3/is_background_set.c | 117 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 meson.build               |   1 +
 src/main.c                |  18 ++++++++++++++++--
 5 files changed, 149 insertions(+), 2 deletions(-)
```

psychon committed:
Apply the diff from clang-format

Signed-off-by: Uli Schlachter <psychon@znc.in>

stapelberg commented on RELEASE-NOTES-next:
```diff
@@ -6,6 +6,13 @@
 This is i3 v4.19. This version is considered stable. All users of i3 are
 strongly encouraged to upgrade.
 
+i3 used to unconditionally take a screenshot and set it as the wallpaper when
```
I would like to make this a little more clear to users. How about:

---

Background/wallpaper workaround:

Some login managers (e.g. gdm) start X11 with the -background none flag. When this flag is set, a background needs to be explicitly set later in the X11 session, otherwise stale copies of closed windows remain visible on the X11 root window (symptom looks like “my terminal window is not closing”).

i3 works around this situation by setting a screenshot of the X11 root window as background when starting. Any background you set before starting i3 (e.g. in your Xsession) or after starting i3 (e.g. via exec statements in the i3 config) will be visible.

A downside of this workaround is that if you have any windows open in your X11 session, those will be part of the screenshot.

To fix this issue, starting in v4.20, i3 detects whether the -background none option is enabled and only then sets a screenshot as background.

---

Notably, the suggestion to add xsetroot shouldn’t be necessary, correct?

Also, in which situations would one typically encounter open X11 windows before starting i3?

psychon replied:
> How about:

Thanks! I applied that in an extra commit did some small changes ontop of it. Since the line wrapping causes that commit to change "everything", here is `git show --word-diff` for that latest commit:
```
Some login managers (e.g. gdm) start {+the+} X11 {+server+} with the -background none
flag. When this flag is set, a background needs to be explicitly set later in
the X11 session, otherwise stale copies of closed windows remain visible on the
X11 root window (symptom looks like “my terminal window is not closing”).

i3 works around this situation by setting a screenshot[-of the X11 root window-] as background when
starting. Any background you set before starting i3 (e.g. in your Xsession) or
after starting i3 (e.g. via exec statements in the i3 config) will be visible.

A downside of this workaround is that if you have any windows {+already+} open in
your X11 session, those will be part of the screenshot.
```

> Notably, the suggestion to add xsetroot shouldn’t be necessary, correct?

Oh, right. That was totally a left over from the previous "just drop the screenshot". Thanks for catching that!

> Also, in which situations would one typically encounter open X11 windows before starting i3?

Uhm... Good question.

Hypothetical scenario: "I used to use fluxbox and for autostart I had things like `xterm& thunderbird& whatever & exec fluxbox` in my `~/.xsession`. I kept that stuff in there when switching to i3".

In the above situation, there is a race and e.g. xterm might be able to get its window visible before the WM starts.

Another idea would be `i3 --replace` replacing an already running WM, but it is very likely that the original WM exiting causes the whole X11 session do be brought down. (But this is only possible after `--replace` actually exists, which is #4374 which I still have to update again...)

stapelberg replied:
Thanks for clarifying! I’m thinking the section is no longer as relevant as it used to be (no action required from users), so I might move it into its own section when preparing the next release. It’s fine to merge it as-is, though, I’d say.

psychon committed:
Incorporate suggested wording from Michael

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon committed:
My own changes to Michael's suggestion

Signed-off-by: Uli Schlachter <psychon@znc.in>
